MFLT-5058 Document the version matrix in the platform OTA doc (#148)
memfault/memfault-docs: docs/platform/ota.mdx

# Over-the-Air Updates (OTA)

> Memfault provides secure and scalable updates for all IoT devices ([MCU][4],

Author: https://memfault.slack.com/archives/C025LBJDQRM/p1641994741018100?thread_ts=1641989008.015200&cid=C025LBJDQRM

> @Heiko the numbering of the reference-style links was introduced because I thought only numbers are supported, after noticing Prettier broke my links at the bottom of the doc.
As it turns out, Prettier just doesn't support those links, and it breaks them whatever you put in there. I may stop using reference-style links...

one To Version. The behavior can be described as: `[1.0→2.0]`.

Note that Delta Releases are referred to as ["incremental updates" in Android

> documentation][2], or "differential updates" in other places.

Author: https://github.com/memfault/memfault-docs/pull/148#discussion_r783050383

the top-right corner of the _OTA payloads_ card.

Alongside the actual payload file, you'll be prompted for different inputs that
a device querying the [Latest Release endpoint][1] must match for this Release
to be applicable to it.

See the [updating behavior](#updating-behavior) section for more details.

### Cohort Version Management

To dive into your OTA configuration for a specific Cohort, you can access the
**Cohort detail** view by first opening the **Cohorts** list view from the main
menu, and then selecting the Cohort you're interested in.

> Under the _Version Management_ tab, you'll find the _version matrix_: a live

Reviewer: Not sure I fully understand the typography system, but if I do: should "Version Management" be bold as it's a UI tab?

Author: Yup, it should :+1:

view into the very next OTA update that each device in this Cohort will receive.

> The version matrix is a table. Its rows represent _current versions_ being
> reported by devices in the field. Its columns represent _future versions_. What

Reviewer: If the matrix is a table, why do we call it matrix and not table?

```suggestion
The version matrix consists of rows representing _current versions_ that are currently reported by devices in the field. Its columns represent _future versions_ devices of the cohort will report (after having applied OTA releases). What
```

follows is a tour of the different interface elements of the deployment matrix.

<figure id="figure-1" style={DEPLOYMENT_MATRIX_FIGURE_STYLES}>
  <img
    src="/img/docs/platform/ota/matrix-focus-full-releases-min.png"
    alt="Figure 1: Full Releases highlighted on the version matrix"
  />
  <figcaption>
    Figure 1: Full Releases highlighted on the version matrix.
  </figcaption>
</figure>

[Figure 1](#figure-1) highlights the area of the version matrix that shows
future versions. In this example, three Full Releases exist (for versions 1.0,
2.0 and 1.1), but only Full Releases 1.0 and 2.0 are active. This is represented
by the blue checkbox icon <CheckSquareTwoTone />. The dotted line around the

> icon for the 2.2 version means that this version has no Full Release for it.

Reviewer: You say "it" does not have a release. Should we -throughout the doc- emphasize that it's not the matrix "having" something but the cohort and that the matrix merely visualizes it?

Author: I agree with your intention but `it` here means the version, not the matrix.

Changed the line to this: `The dotted line around the icon for the 2.2 version means that there is no Full Release with the same version.`. It repeats the word "version" but is a lot clearer.

The <TargetVersionFilled /> icon (which conventionally means "final state") on
top of the 2.2 version shows us that 2.2 is a final target version for this

> Cohort. If all devices in this Cohort were to take all available Releases until
> no further Release is available, then a Release to version 2.2 would be the last
> one they would take.

Reviewer: ```suggestion
Cohort. Eventually, all devices in this cohort will end up on this version 2.2.
```

<figure id="figure-2" style={DEPLOYMENT_MATRIX_FIGURE_STYLES}>
  <img
    src="/img/docs/platform/ota/matrix-focus-delta-releases-min.png"
    alt="Figure 2: Delta Releases highlighted on the version matrix"
  />
  <figcaption>
    Figure 2: Delta Releases highlighted on the version matrix.
  </figcaption>
</figure>

[Figure 2](#figure-2) highlights the area of the version matrix that represents

> Delta Releases. The _Current Version_ column is where you may see the From

Reviewer: should we re-iterate what a delta release is or link to the description above? Same for full releases?

Version of a Delta Release, and the _Target_ column is where the Target Version

> of a Delta Release can be found. A drop-down icon without a version represents a
> Software Version reported by devices that does not have a matching Delta
> Release. In the picture, a Delta Release from 2.0 to 2.1 is active (marked with

Reviewer: ```suggestion
of a Delta Release can be found. An empty drop-down icon indicates that there is no active Delta Release from the row's version. In the picture, only a Delta Release from 2.0 to 2.1 is active (marked with
```

a <CheckSquareTwoTone />).

<figure id="figure-3" style={DEPLOYMENT_MATRIX_FIGURE_STYLES}>
  <img
    src="/img/docs/platform/ota/matrix-full-and-delta-min.png"
    alt="Figure 3: Full Releases and Delta Releases on the version matrix"
  />
  <figcaption>
    Figure 3: Full Releases and Delta Releases on the version matrix.
  </figcaption>
</figure>

> [Figure 3](#figure-3) shows the following scenario:

Reviewer: I am not sure we ever said before, what the cells in the matrix represent.

Author: It's right after this:

```
- 21K devices currently on 0.9 are going to download a Full Release to 2.0.
- 36K devices currently on 1.1 are going to download a Full Release to 2.0.
- 132K devices currently on 2.0 are going to download a Delta Release from 2.0
  to 2.1.
- 738K devices are already on 2.1 and there's no active Release available to
  them.
```

Maybe I should give a generic introduction before the examples.

Author: ```
In each cell of the version matrix you'll find a count of devices. If, for
example, a device count of 10K showed up on a cell in the intersection of row
version 1.1 and the column 2.0, that would mean 10K devices are currently on
version 1.1 and the immediate next step they're going to take is to download an
OTA payload for a Release with version 2.0.

Conversely, if 5K devices were to show up on a cell in a row with version 1.0
and under the column for version 1.0, that would mean those 5K devices will stay
on version 1.0.

All device counts on the version matrix are clickable, and will take you to a
complete list of the devices represented under that count.
```

- Two active Full Releases exist (1.0 and 2.0) and there's one active Delta
  Release from 2.0 to 2.1.

> - Full Release 1.0 is currently not the target of any devices.

Reviewer: "target" is misleading here. Also nit: Wouldn't we filter out that column in practice?

Reviewer: It's also worth calling out that the matrix at any time describes (only) the *next* step that applies to the devices.

Author: Changed the sentence to `Full Release 1.0 is currently not applicable to any device.`

- 21K devices currently on 0.9 are going to download a Full Release to 2.0.
- 36K devices currently on 1.1 are going to download a Full Release to 2.0.
- 132K devices currently on 2.0 are going to download a Delta Release from 2.0
  to 2.1.

> - 738K devices are already on 2.1 and there's no active Release available to

Reviewer: ```suggestion
- 738K devices are already on 2.1. They will not receive any updates and stay at 2.1
```

  them.

<figure id="figure-4" style={DEPLOYMENT_MATRIX_FIGURE_STYLES}>
  <img
    src="/img/docs/platform/ota/matrix-staged-min.png"
    alt="Figure 4: A staged rollout on the version matrix"
  />
  <figcaption>Figure 4: a staged rollout on the version matrix.</figcaption>
</figure>

[Figure 4](#figure-4) shows the addition of a
[staged rollout](#staged-rollout-of-a-release) of a Delta Release from version
2.1 to version 2.2. The percentage icon <StyledPercentageOutlined /> indicates a
staged rollout. In this scenario, only 5.9K devices out of a total of 54K that
are on version 2.1 will download the payload from Delta Release from 2.1 to 2.2.

> #### Managing OTA Updates for a specific Cohort

Reviewer: should we scratch the "for a specific Cohort" from the headline? When scanning the outline, it may suggest that everything else is not limited to a single cohort.

The version matrix in the Cohort Version Management tab can also be used to
manage your OTA Releases.

<figure id="figure-5" style={DEPLOYMENT_MATRIX_FIGURE_STYLES}>
  <img
    src="/img/docs/platform/ota/matrix-actions-min.png"
    alt="Figure 5: Highlighted actions on the version matrix"
  />
  <figcaption>Figure 5: highlighted actions on the version matrix.</figcaption>
</figure>

[Figure 5](#figure-5) highlights the location of buttons on the version matrix
that allow for fine-grained control of OTA Releases in a specific Cohort.

You can manage Full Releases by using the controls arranged horizontally on the
top side of the version matrix:

- Click on the plus icon on the right end to create a new Full Release.
- Use the highlighted buttons under each future version to create a Full Release

>   targeting this version, or to create a Release Activation, abort it, or
>   configure it.

Reviewer: I'd avoid the artificial term "Release Activation".

```suggestion
  targeting this version, or to abort or configure active Releases.
```

Author: It's in the UI now as well. I think in some instances it was necessary to avoid awkward phrasing. But we may want to review this.

the top-right corner of the _OTA payloads_ card.

Alongside the actual payload file, you'll be prompted for different inputs that
a device querying the [Latest Release endpoint][1] must match for this Release
to be applicable to it.

See the [updating behavior](#updating-behavior) section for more details.

### Cohort Version Management

To dive into your OTA configuration for a specific Cohort, you can access the

> **Cohort detail** view by first opening the **Cohorts** list view from the main

Reviewer 2: OT: did we establish "detail" as a term? Is this here establishing it as a term? When are we using `_` vs `**`?

Author: At least in my documents:

-  `_` is for normal prose-like emphasis
- `**` is to denote a view in the app (we could use a component for this)

You created a "How to write copy" document, maybe we should also start a "How to write documentation" doc?

About "detail": we use it elsewhere in the docs, but we've never talked about it. Ditto on the doc.

menu, and then selecting the Cohort you're interested in.

> Under the _Version Management_ tab, you'll find what we like to call the

Reviewer 2: ```suggestion
Under the _Version Management_ tab, you'll find the
```

_version matrix_: a live view into the very next OTA update that each device in

> this Cohort will receive (if any).

Reviewer 2: ```suggestion
this Cohort will receive.
```

> The version matrix looks like a table. Its rows represent _current versions_

Reviewer 2: ```suggestion
The version matrix is a table. Its rows represent _current versions_
```

  />
  <figcaption>
    Figure 1: Full Releases highlighted on the version matrix.
  </figcaption>
</figure>

[Figure 1](#figure-1) highlights the area of the version matrix that shows
future versions. In this example, three Full Releases exist (for versions 1.0,
2.0 and 1.1), but only Full Releases 1.0 and 2.0 are active. This is represented
by the blue checkbox icon <CheckSquareTwoTone />. The dotted line around the
icon for the 2.2 version means that this version has no Full Release for it.

The <TargetVersionFilled /> icon (which conventionally means "final state") on
top of the 2.2 version shows us that 2.2 is a final target version for this
Cohort. If all devices in this Cohort were to take all available Releases until
no further Release is available, then a Release to version 2.2 would be the last
one they would take.

<figure id="figure-2" style={DEPLOYMENT_MATRIX_FIGURE_STYLES}>
  <img
    src="/img/docs/platform/ota/matrix-focus-delta-releases-min.png"
    alt="Figure 2: Delta Releases highlighted on the version matrix"
  />
  <figcaption>
    Figure 2: Delta Releases highlighted on the version matrix.
  </figcaption>
</figure>

> [Figure 2](#figure-2) highlights the are of the version matrix that represents

Reviewer 2: Broken sentence? Not sure what this is supposed to read like.

Author: Yup good catch, probably a bad merge since I had a nasty conflict.

Author: Actually a typo: `highlights the area`
